File: src/main/java/com/lessmarkup/engine/data/migrate/MigrateEngine.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.lessmarkup.engine.data.migrate;

import com.lessmarkup.dataobjects.MigrationHistory;
import com.lessmarkup.framework.helpers.StringHelper;
import com.lessmarkup.framework.system.RequestContextHolder;
import com.lessmarkup.interfaces.data.DomainModel;
import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.data.Migration;
import com.lessmarkup.interfaces.data.Migrator;
import com.lessmarkup.interfaces.exceptions.CommonException;
import com.lessmarkup.interfaces.module.ModuleConfiguration;
import com.lessmarkup.interfaces.module.ModuleProvider;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.time.OffsetDateTime;
import java.util.HashSet;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

@Component
@Scope("prototype")
public class MigrateEngine {
    
    private final ModuleProvider moduleProvider;
    private final DomainModelProvider domainModelProvider;

    @Autowired
    public MigrateEngine(ModuleProvider moduleProvider, DomainModelProvider domainModelProvider) {
        this.moduleProvider = moduleProvider;
        this.domainModelProvider = domainModelProvider;
    }

    public void execute() {
        String connectionString = RequestContextHolder.getContext().getEngineConfiguration().getConnectionString();
        if (StringHelper.isNullOrEmpty(connectionString)) {
            return;
        }
        
        execute(connectionString);
    }
    
    public void execute(String connectionString) {
        
        if (StringHelper.isNullOrEmpty(connectionString)) {
            throw new CommonException("Connection string is empty");
        }
        
        Migrator migrator = new MigratorImpl(connectionString);
        
        if (!migrator.checkExists(MigrationHistory.class)) {
            migrator.createTable(MigrationHistory.class);
        }
        
        try (DomainModel domainModel = domainModelProvider.create(connectionString)) {
            Set<String> existingMigrations = new HashSet<>();
            domainModel.query()
                    .from(MigrationHistory.class)
                    .toList(MigrationHistory.class, "uniqueId")
                    .forEach(h -> existingMigrations.add(h.getUniqueId()));
            
            for (ModuleConfiguration module : moduleProvider.getModules()) {
                for (Class<? extends Migration> migrationType : module.getInitializer().getMigrations()) {
                    try {
                        Migration migration = migrationType.newInstance();
                        String uniqueId = migration.getId() + "_" + migrationType.getName();
                        if (existingMigrations.contains(uniqueId)) {
                            continue;
                        }
                        migration.migrate(migrator);
                        MigrationHistory history = new MigrationHistory();
                        history.setCreated(OffsetDateTime.now());
                        history.setUniqueId(uniqueId);
                        history.setModuleType(module.getModuleType());
                        domainModel.create(history);
                    } catch (InstantiationException | IllegalAccessException ex) {
                        Logger.getLogger(MigrateEngine.class.getName()).log(Level.SEVERE, null, ex);
                    }
                }
            }
        }
    }
}

File: src/main/scala/com/lessmarkup/engine/data/migrate/MigrateEngine.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.lessmarkup.engine.data.migrate

import java.time.OffsetDateTime
import java.util.logging.{Level, Logger}

import com.google.inject.Inject
import com.lessmarkup.dataobjects.MigrationHistory
import com.lessmarkup.framework.helpers.StringHelper
import com.lessmarkup.framework.system.RequestContextHolder
import com.lessmarkup.interfaces.data.{DomainModel, DomainModelProvider, Migration, Migrator}
import com.lessmarkup.interfaces.exceptions.CommonException
import com.lessmarkup.interfaces.module.ModuleProvider

class MigrateEngine @Inject() (moduleProvider: ModuleProvider, domainModelProvider: DomainModelProvider) {

  def execute() {
    val connectionString: String = RequestContextHolder.getContext.getEngineConfiguration.getConnectionString
    if (StringHelper.isNullOrEmpty(connectionString)) {
      return
    }
    execute(connectionString)
  }

  def execute(connectionString: String) {
    if (StringHelper.isNullOrEmpty(connectionString)) {
      throw new CommonException("Connection string is empty")
    }
    val migrator: Migrator = new MigratorImpl(connectionString)
    if (!migrator.checkExists(classOf[MigrationHistory])) {
      migrator.createTable(classOf[MigrationHistory])
    }
    val domainModel: DomainModel = domainModelProvider.create(connectionString)
    try {
      val existingMigrations = domainModel.query
        .from(classOf[MigrationHistory])
        .toList(classOf[MigrationHistory], Option("uniqueId"))
        .map(_.uniqueId)
        .toSet
      for (module <- moduleProvider.getModules) {
        for (migrationType <- module.getInitializer.getMigrations) {
          try {
            val migration: Migration = migrationType.newInstance
            val uniqueId = migration.getId + "_" + migrationType.getName
            if (!existingMigrations.contains(uniqueId)) {
              migration.migrate(migrator)
              val history: MigrationHistory = new MigrationHistory(
                created = OffsetDateTime.now,
                uniqueId = uniqueId,
                moduleType = module.getModuleType,
                id = 0
              )
              domainModel.create(history)
            }
          }
          catch {
            case ex: Any =>
              Logger.getLogger(classOf[MigrateEngine].getName).log(Level.SEVERE, null, ex)
          }
        }
      }
    } finally {
      if (domainModel != null) domainModel.close()
    }
  }
}